src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs

```diff
@@ -2395,24 +2426,68 @@ private static void AddNestedTypesToDictionary(Dictionary<string, ImmutableArray
         {
-            public ArrayBuilder<Symbol> NonTypeNonIndexerMembers { get; set; } = ArrayBuilder<Symbol>.GetInstance();
-            public readonly ArrayBuilder<ArrayBuilder<FieldOrPropertyInitializer.Builder>> StaticInitializers = ArrayBuilder<ArrayBuilder<FieldOrPropertyInitializer.Builder>>.GetInstance();
-            public readonly ArrayBuilder<ArrayBuilder<FieldOrPropertyInitializer.Builder>> InstanceInitializers = ArrayBuilder<ArrayBuilder<FieldOrPropertyInitializer.Builder>>.GetInstance();
-            public readonly ArrayBuilder<SyntaxReference> IndexerDeclarations = ArrayBuilder<SyntaxReference>.GetInstance();
+            public ArrayBuilder<Symbol> NonTypeNonIndexerMembers { get; set; }
+            public readonly ArrayBuilder<ArrayBuilder<FieldOrPropertyInitializer.Builder>> StaticInitializers;
+            public readonly ArrayBuilder<ArrayBuilder<FieldOrPropertyInitializer.Builder>> InstanceInitializers;
+            public readonly ArrayBuilder<SyntaxReference> IndexerDeclarations;
             public RecordDeclarationSyntax? RecordDeclarationWithParameters;
-            public ArrayBuilder<FieldOrPropertyInitializer.Builder>? InstanceInitializersForRecordDeclarationWithParameters;
+            public int InstanceInitializersIndexForRecordDeclarationWithParameters;
             public bool IsNullableEnabledForInstanceConstructorsAndFields;
             public bool IsNullableEnabledForStaticConstructorsAndFields;
 
-            public MembersAndInitializers ToReadOnlyAndFree()
+            public MembersAndInitializersBuilder()
+            {
+                NonTypeNonIndexerMembers = new ArrayBuilder<Symbol>();
+                StaticInitializers = new ArrayBuilder<ArrayBuilder<FieldOrPropertyInitializer.Builder>>();
+                InstanceInitializers = new ArrayBuilder<ArrayBuilder<FieldOrPropertyInitializer.Builder>>();
+                IndexerDeclarations = new ArrayBuilder<SyntaxReference>();
+                InstanceInitializersIndexForRecordDeclarationWithParameters = -1;
+            }
+
+            private MembersAndInitializersBuilder(MembersAndInitializersBuilder original)
+            {
+                NonTypeNonIndexerMembers = new ArrayBuilder<Symbol>(original.NonTypeNonIndexerMembers.Count);
+                NonTypeNonIndexerMembers.AddRange(original.NonTypeNonIndexerMembers);
+
+                StaticInitializers = new ArrayBuilder<ArrayBuilder<FieldOrPropertyInitializer.Builder>>(original.StaticInitializers.Count);
+                foreach (var staticInitializers in original.StaticInitializers)
+                {
+                    var copy = new ArrayBuilder<FieldOrPropertyInitializer.Builder>(staticInitializers.Count);
+                    copy.AddRange(staticInitializers);
+                    StaticInitializers.Add(copy);
+                }
+
+                InstanceInitializers = new ArrayBuilder<ArrayBuilder<FieldOrPropertyInitializer.Builder>>(original.InstanceInitializers.Count);
```

Comment:
>InstanceInitializers = new ArrayBuilder<ArrayBuilder<FieldOrPropertyInitializer.Builder>>(original.InstanceInitializers.Count); [](start = 16, length = 127)

Consider sharing code with the previous section that deals with static initializers.